File: android/core-protocol/src/main/kotlin/app/candela/protocol/Fountain.kt
package app.candela.protocol

import java.util.BitSet
import kotlin.math.ceil
import kotlin.math.ln
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Systematic Luby Transform fountain. 1:1 with src/protocol/fountain.ts.
 *
 * THIS FILE IS INTEROP-CRITICAL. The neighbour selection is a pure function of
 * (symbolId, k) and must produce byte-identical results to the TypeScript
 * reference, or a Kotlin receiver silently mis-XORs a TypeScript sender's
 * symbols and the fountain never converges. FountainNeighborsTest asserts this
 * against golden vectors for k in {1, 2, 7, 86, 683, 2000} across ~5200 ids.
 *
 * The two places JS and JVM most easily diverge, both handled explicitly below:
 *  1. mulberry32 — JS uses uint32 wraparound and Math.imul; Kotlin Int is signed
 *     32-bit with the same wraparound, so `imul` maps to plain `*`, but every
 *     shift must be `ushr` (logical) and every comparison/division must go
 *     through a Long widened with `and 0xFFFFFFFFL`.
 *  2. seeding — `(symbolId + 1) * 0x9E3779B9` is IEEE-754 double arithmetic in
 *     JS, exact for ids up to 65535 (product < 2^53), then truncated to uint32.
 *     Kotlin reproduces that with Long multiply + mask.
 */
object Fountain {

    /** Deterministic PRNG, bit-exact with the TypeScript mulberry32. */
    class Mulberry32(seed: Int) {
        private var a: Int = seed

        fun nextDouble(): Double {
            a += 0x6D2B79F5.toInt()
            var t = a
            t = (t xor (t ushr 15)) * (t or 1)
            t = t xor (t + (t xor (t ushr 7)) * (t or 61))
            val u = (t xor (t ushr 14)).toLong() and 0xFFFFFFFFL
            return u.toDouble() / 4294967296.0
        }
    }

    fun seedFor(symbolId: Int): Int =
        (((symbolId + 1).toLong() * 0x9E3779B9L) and 0xFFFFFFFFL).toInt()

    /**
     * Robust soliton distribution. Returns mu[0..k], mu[0] unused.
     * c = 0.12, delta = 0.05 (frozen, PSR section 2.4).
     */
    fun robustSoliton(
        k: Int,
        c: Double = Constants.FOUNTAIN_C,
        delta: Double = Constants.FOUNTAIN_DELTA,
    ): DoubleArray {
        val r = max(1.0, c * ln(k / delta) * sqrt(k.toDouble()))
        val rho = DoubleArray(k + 1)
        if (k >= 1) rho[1] = 1.0 / k
        for (d in 2..k) rho[d] = 1.0 / (d.toDouble() * (d - 1).toDouble())

        val spike = max(1, min(k, Math.round(k / r).toInt()))
        val tau = DoubleArray(k + 1)
        for (d in 1 until spike) tau[d] = r / (d.toDouble() * k.toDouble())
        if (spike <= k) tau[spike] = (r * ln(r / delta)) / k.toDouble()

        var z = 0.0
        val mu = DoubleArray(k + 1)
        for (d in 1..k) {
            mu[d] = rho[d] + tau[d]
            z += mu[d]
        }
        for (d in 1..k) mu[d] /= z
        return mu
    }

    private fun sampleDegree(mu: DoubleArray, rng: Mulberry32): Int {
        val u = rng.nextDouble()
        var acc = 0.0
        for (d in 1 until mu.size) {
            acc += mu[d]
            if (u <= acc) return d
        }
        return max(1, mu.size - 1)
    }

    private fun sampleNeighbors(k: Int, d: Int, rng: Mulberry32): IntArray {
        val set = LinkedHashSet<Int>()
        val want = min(max(1, d), k)
        var guard = 0
        while (set.size < want && guard++ < k * 8) {
            val idx = Math.floor(rng.nextDouble() * k).toInt()
            if (idx in 0 until k) set.add(idx)
        }
        val out = set.toIntArray()
        out.sort()
        return out
    }

    /** Systematic: ids below k are the source blocks themselves. */
    fun neighborsFor(symbolId: Int, k: Int, mu: DoubleArray): IntArray {
        if (symbolId < k) return intArrayOf(symbolId)
        val rng = Mulberry32(seedFor(symbolId))
        val d = sampleDegree(mu, rng)
        return sampleNeighbors(k, d, rng)
    }

    internal fun xorInto(dst: ByteArray, src: ByteArray) {
        val n = min(dst.size, src.size)
        for (i in 0 until n) dst[i] = (dst[i].toInt() xor src[i].toInt()).toByte()
    }

    class Encoder(data: ByteArray, val blockSize: Int) {
        val fileSize: Int = data.size
        val k: Int = max(1, ceil(data.size.toDouble() / blockSize).toInt())
        val mu: DoubleArray
        private val blocks: Array<ByteArray>
        private var nextId = 0

        init {
            blocks = Array(k) { i ->
                val b = ByteArray(blockSize)
                val start = i * blockSize
                val end = min(start + blockSize, data.size)
                if (start < data.size) data.copyInto(b, 0, start, end)
                b
            }
            mu = robustSoliton(k)
        }

        fun recommendedSymbols(): Int =
            ceil(k * Constants.FOUNTAIN_OVERHEAD).toInt() + 16

        data class Symbol(val symbolId: Int, val payload: ByteArray, val neighbors: IntArray) {
            override fun equals(other: Any?) = other is Symbol &&
                symbolId == other.symbolId && payload.contentEquals(other.payload)
            override fun hashCode() = symbolId
        }

        fun encode(symbolId: Int? = null): Symbol {
            val id = symbolId ?: nextId++
            val neighbors = neighborsFor(id, k, mu)
            val payload = ByteArray(blockSize)
            for (n in neighbors) xorInto(payload, blocks[n])
            return Symbol(id, payload, neighbors)
        }
    }

    private class Equation(var vars: MutableList<Int>, val payload: ByteArray)

    /**
     * Peel decoder with a bounded Gaussian-elimination fallback.
     *
     * Peeling alone stalls on mid-size k; GE finishes it. GE is bounded to
     * <=80 unknowns and a pool of <=96 equations so worst-case cost stays inside
     * the thermal budget (audit section 3) rather than spiking on a bad frame.
     *
     * Dedup uses a BitSet(65536) — allocation-free after init, matching the
     * 16-bit symbol id space (audit section 6.6).
     */
    class Decoder(val k: Int, val blockSize: Int, val fileSize: Int) {
        val mu: DoubleArray = robustSoliton(k)
        private val recovered = arrayOfNulls<ByteArray>(k)
        private var recoveredCount = 0
        private val seen = BitSet(65536)
        private var seenCount = 0
        private var pool = mutableListOf<Equation>()

        val uniqueCount: Int get() = seenCount
        val doneCount: Int get() = recoveredCount
        val progress: Double get() = if (k == 0) 1.0 else recoveredCount.toDouble() / k

        fun has(symbolId: Int): Boolean = seen.get(symbolId)
        fun isComplete(): Boolean = recoveredCount >= k

        fun ingest(symbolId: Int, payload: ByteArray): Boolean {
            if (symbolId < 0 || symbolId > 65535) return false
            if (seen.get(symbolId) || isComplete()) return false
            seen.set(symbolId)
            seenCount++

            val neighbors = neighborsFor(symbolId, k, mu)
            val eq = Equation(neighbors.toMutableList(), payload.copyOf())
            substituteKnown(eq)
            pool.add(eq)
            peel()
            if (!isComplete() && pool.isNotEmpty() && pool.size <= 96) tryGaussian()
            return true
        }

        private fun substituteKnown(eq: Equation) {
            var i = eq.vars.size - 1
            while (i >= 0) {
                val n = eq.vars[i]
                val rec = recovered[n]
                if (rec != null) {
                    xorInto(eq.payload, rec)
                    eq.vars.removeAt(i)
                }
                i--
            }
        }

        private fun peel() {
            var progressed = true
            while (progressed) {
                progressed = false
                var i = pool.size - 1
                while (i >= 0) {
                    val eq = pool[i]
                    substituteKnown(eq)
                    if (eq.vars.isEmpty()) {
                        pool.removeAt(i)
                    } else if (eq.vars.size == 1) {
                        val idx = eq.vars[0]
                        if (recovered[idx] == null) {
                            recovered[idx] = eq.payload.copyOf()
                            recoveredCount++
                            progressed = true
                        }
                        pool.removeAt(i)
                    }
                    i--
                }
            }
        }

        private fun tryGaussian() {
            val unknown = ArrayList<Int>()
            for (i in 0 until k) if (recovered[i] == null) unknown.add(i)
            if (unknown.isEmpty() || unknown.size > 80) return

            val col = HashMap<Int, Int>(unknown.size * 2)
            unknown.forEachIndexed { i, id -> col[id] = i }
            val u = unknown.size

            val eqs = ArrayList<Equation>()
            for (eq in pool) {
                substituteKnown(eq)
                if (eq.vars.isEmpty()) continue
                if (eq.vars.all { col.containsKey(it) }) {
                    eqs.add(Equation(eq.vars.toMutableList(), eq.payload.copyOf()))
                }
            }
            if (eqs.size < u) return

            val rowW = (u + 7) / 8
            val bits = ArrayList<ByteArray>(eqs.size)
            val rhs = ArrayList<ByteArray>(eqs.size)
            for (eq in eqs) {
                val row = ByteArray(rowW)
                for (v in eq.vars) {
                    val c = col[v] ?: continue
                    row[c shr 3] = (row[c shr 3].toInt() or (1 shl (c and 7))).toByte()
                }
                bits.add(row)
                rhs.add(eq.payload.copyOf())
            }

            val m = bits.size
            val pivotRow = IntArray(u) { -1 }
            var rank = 0
            for (c in 0 until u) {
                var pr = -1
                for (r in rank until m) {
                    if ((bits[r][c shr 3].toInt() and (1 shl (c and 7))) != 0) { pr = r; break }
                }
                if (pr < 0) continue
                if (pr != rank) {
                    val tb = bits[rank]; bits[rank] = bits[pr]; bits[pr] = tb
                    val tp = rhs[rank]; rhs[rank] = rhs[pr]; rhs[pr] = tp
                }
                pivotRow[c] = rank
                for (r in 0 until m) {
                    if (r == rank) continue
                    if ((bits[r][c shr 3].toInt() and (1 shl (c and 7))) != 0) {
                        val br = bits[r]; val bk = bits[rank]
                        for (j in 0 until rowW) br[j] = (br[j].toInt() xor bk[j].toInt()).toByte()
                        xorInto(rhs[r], rhs[rank])
                    }
                }
                rank++
            }
            if (rank < u) return

            for (c in 0 until u) {
                val r = pivotRow[c]
                if (r < 0) return
                val idx = unknown[c]
                if (recovered[idx] == null) {
                    recovered[idx] = rhs[r].copyOf()
                    recoveredCount++
                }
            }
            pool = mutableListOf()
        }

        fun assemble(): ByteArray? {
            if (!isComplete()) return null
            val out = ByteArray(fileSize)
            for (i in 0 until k) {
                val block = recovered[i] ?: return null
                val start = i * blockSize
                val len = min(blockSize, fileSize - start)
                if (len > 0) block.copyInto(out, start, 0, len)
            }
            return out
        }
    }
}

File: src/protocol/fountain.ts
import { FOUNTAIN_C, FOUNTAIN_DELTA, FOUNTAIN_OVERHEAD } from "./constants";

function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function robustSoliton(k: number, c = FOUNTAIN_C, delta = FOUNTAIN_DELTA): Float64Array {
  const R = Math.max(1, c * Math.log(k / delta) * Math.sqrt(k));
  const rho = new Float64Array(k + 1);
  rho[1] = 1 / k;
  for (let d = 2; d <= k; d++) rho[d] = 1 / (d * (d - 1));

  const spike = Math.max(1, Math.min(k, Math.round(k / R)));
  const tau = new Float64Array(k + 1);
  for (let d = 1; d < spike; d++) tau[d] = R / (d * k);
  tau[spike] = (R * Math.log(R / delta)) / k;

  let z = 0;
  const mu = new Float64Array(k + 1);
  for (let d = 1; d <= k; d++) {
    mu[d] = rho[d]! + tau[d]!;
    z += mu[d]!;
  }
  for (let d = 1; d <= k; d++) mu[d]! /= z;
  return mu;
}

function sampleDegree(mu: Float64Array, rng: () => number): number {
  const u = rng();
  let acc = 0;
  for (let d = 1; d < mu.length; d++) {
    acc += mu[d]!;
    if (u <= acc) return d;
  }
  return Math.max(1, mu.length - 1);
}

function sampleNeighbors(k: number, d: number, rng: () => number): number[] {
  const set = new Set<number>();
  const want = Math.min(Math.max(1, d), k);
  let guard = 0;
  while (set.size < want && guard++ < k * 8) {
    const idx = Math.floor(rng() * k);
    if (idx >= 0 && idx < k) set.add(idx);
  }
  return Array.from(set).sort((a, b) => a - b);
}

export function neighborsFor(symbolId: number, k: number, mu: Float64Array): number[] {
  if (symbolId < k) return [symbolId];
  const rng = mulberry32((symbolId + 1) * 0x9e3779b9);
  const d = sampleDegree(mu, rng);
  return sampleNeighbors(k, d, rng);
}

function xorInto(dst: Uint8Array, src: Uint8Array): void {
  const n = Math.min(dst.length, src.length);
  for (let i = 0; i < n; i++) dst[i]! ^= src[i]!;
}

type Equation = {
  vars: number[];
  payload: Uint8Array;
};

export class FountainEncoder {
  readonly k: number;
  readonly blockSize: number;
  readonly blocks: Uint8Array[];
  readonly fileSize: number;
  readonly mu: Float64Array;
  private nextId = 0;

  constructor(data: Uint8Array, blockSize: number) {
    this.fileSize = data.length;
    this.blockSize = blockSize;
    this.k = Math.max(1, Math.ceil(data.length / blockSize));
    this.blocks = [];
    for (let i = 0; i < this.k; i++) {
      const b = new Uint8Array(blockSize);
      const start = i * blockSize;
      b.set(data.subarray(start, Math.min(start + blockSize, data.length)));
      this.blocks.push(b);
    }
    this.mu = robustSoliton(this.k);
  }

  recommendedSymbols(): number {
    return Math.ceil(this.k * FOUNTAIN_OVERHEAD) + 16;
  }

  encode(symbolId?: number): { symbolId: number; payload: Uint8Array; neighbors: number[] } {
    const id = symbolId ?? this.nextId++;
    const neighbors = neighborsFor(id, this.k, this.mu);
    const payload = new Uint8Array(this.blockSize);
    for (const n of neighbors) xorInto(payload, this.blocks[n]!);
    return { symbolId: id, payload, neighbors };
  }
}

export class FountainDecoder {
  readonly k: number;
  readonly blockSize: number;
  readonly fileSize: number;
  readonly mu: Float64Array;
  private recovered: (Uint8Array | null)[];
  private recoveredCount = 0;
  private seen = new Set<number>();
  private pool: Equation[] = [];

  constructor(k: number, blockSize: number, fileSize: number) {
    this.k = k;
    this.blockSize = blockSize;
    this.fileSize = fileSize;
    this.mu = robustSoliton(k);
    this.recovered = Array.from({ length: k }, () => null);
  }

  get uniqueCount(): number {
    return this.seen.size;
  }

  get doneCount(): number {
    return this.recoveredCount;
  }

  get progress(): number {
    return this.k === 0 ? 1 : this.recoveredCount / this.k;
  }

  has(symbolId: number): boolean {
    return this.seen.has(symbolId);
  }

  ingest(symbolId: number, payload: Uint8Array): boolean {
    if (this.seen.has(symbolId) || this.isComplete()) return false;
    this.seen.add(symbolId);
    const neighbors = neighborsFor(symbolId, this.k, this.mu);
    const eq: Equation = { vars: neighbors.slice(), payload: payload.slice() };
    this.substituteKnown(eq);
    this.pool.push(eq);
    this.peel();
    if (!this.isComplete() && this.pool.length > 0 && this.pool.length <= 96) {
      this.tryGaussian();
    }
    return true;
  }

  private substituteKnown(eq: Equation): void {
    for (let i = eq.vars.length - 1; i >= 0; i--) {
      const n = eq.vars[i]!;
      const rec = this.recovered[n];
      if (rec) {
        xorInto(eq.payload, rec);
        eq.vars.splice(i, 1);
      }
    }
  }

  private peel(): void {
    let progressed = true;
    while (progressed) {
      progressed = false;
      for (let i = this.pool.length - 1; i >= 0; i--) {
        const eq = this.pool[i]!;
        this.substituteKnown(eq);
        if (eq.vars.length === 0) {
          this.pool.splice(i, 1);
        } else if (eq.vars.length === 1) {
          const idx = eq.vars[0]!;
          if (!this.recovered[idx]) {
            this.recovered[idx] = eq.payload.slice();
            this.recoveredCount++;
            progressed = true;
          }
          this.pool.splice(i, 1);
        }
      }
    }
  }

  private tryGaussian(): void {
    const unknown: number[] = [];
    for (let i = 0; i < this.k; i++) if (!this.recovered[i]) unknown.push(i);
    if (unknown.length === 0 || unknown.length > 80) return;
    const col = new Map<number, number>();
    unknown.forEach((id, i) => col.set(id, i));
    const u = unknown.length;
    const eqs: Equation[] = [];
    for (const eq of this.pool) {
      this.substituteKnown(eq);
      if (eq.vars.length === 0) continue;
      if (eq.vars.every((v) => col.has(v))) eqs.push({ vars: eq.vars.slice(), payload: eq.payload.slice() });
    }
    if (eqs.length < u) return;

    const bits: Uint8Array[] = [];
    const rhs: Uint8Array[] = [];
    const rowW = Math.ceil(u / 8);
    for (const eq of eqs) {
      const row = new Uint8Array(rowW);
      for (const v of eq.vars) {
        const c = col.get(v);
        if (c === undefined) continue;
        row[c >> 3]! |= 1 << (c & 7);
      }
      bits.push(row);
      rhs.push(eq.payload.slice());
    }

    const m = bits.length;
    const pivotRow = new Int32Array(u).fill(-1);
    let rank = 0;
    for (let c = 0; c < u; c++) {
      let pr = -1;
      for (let r = rank; r < m; r++) {
        if (bits[r]![c >> 3]! & (1 << (c & 7))) {
          pr = r;
          break;
        }
      }
      if (pr < 0) continue;
      if (pr !== rank) {
        const tb = bits[rank]!;
        bits[rank] = bits[pr]!;
        bits[pr] = tb;
        const tp = rhs[rank]!;
        rhs[rank] = rhs[pr]!;
        rhs[pr] = tp;
      }
      pivotRow[c] = rank;
      for (let r = 0; r < m; r++) {
        if (r === rank) continue;
        if (bits[r]![c >> 3]! & (1 << (c & 7))) {
          for (let j = 0; j < rowW; j++) bits[r]![j]! ^= bits[rank]![j]!;
          xorInto(rhs[r]!, rhs[rank]!);
        }
      }
      rank++;
    }
    if (rank < u) return;

    for (let c = 0; c < u; c++) {
      const r = pivotRow[c] ?? -1;
      if (r < 0) return;
      const idx = unknown[c]!;
      const solved = rhs[r];
      if (!solved) return;
      if (!this.recovered[idx]) {
        this.recovered[idx] = solved.slice();
        this.recoveredCount++;
      }
    }
    this.pool = [];
  }

  isComplete(): boolean {
    return this.recoveredCount >= this.k;
  }

  assemble(): Uint8Array | null {
    if (!this.isComplete()) return null;
    const out = new Uint8Array(this.fileSize);
    for (let i = 0; i < this.k; i++) {
      const block = this.recovered[i];
      if (!block) return null;
      const start = i * this.blockSize;
      const len = Math.min(this.blockSize, this.fileSize - start);
      out.set(block.subarray(0, len), start);
    }
    return out;
  }
}
